### PR TITLE
Refactor Connection Status Logic

### PR DESCRIPTION
**Pull request by williamsyang-work**
This is to improve the tracking of TraceServer status state in the application.  This will support the implementation of the remote use case in the vscode-trace-extension.

This change consists of four commits:

1) **Rename subscription events in Connection client**
'addConnectionStatusListener' changed to 'activate'
'removeConnectionStatusListener' changed to 'deactivate'
This is more accurate.

2) **Add Server Status Change Listeners**
Adds methods to add and remove functions that are executed when the
server status is updated.

3) **Remove Direct DOM Manipulation from ServerStatusWidget**
Removes direct DOM manipulation from TraceExplorerServerStatusWidget.
Refactors it to use the server status change event listeners and change
what is rendered through Theia ReactWidget's component lifecycle.

4) **Check Server Status On Update**
Check the server status every update in the trace-explorer-widget.
This ensures that the 'Open Trace' placeholder widget is displayed
when the server is offline, not just when there are no opened traces.

Signed-off-by: Will Yang <william.yang@ericsson.com>

### CODE REVIEW

File: theia-extensions/viewer-prototype/src/browser/trace-server-connection-status-client-impl.ts
Comment (MatthewKhouzam) [inline theia-extensions/viewer-prototype/src/browser/trace-server-connection-status-client-impl.ts:15]: please rename `fn` all the way down to `listener` or something. This would make the individual functions easier to read.

**Review by MatthewKhouzam — COMMENTED** [review]
Minor nit.

**Comment by williamsyang-work** [comment]
@sgraband I'm proposing changes to the `TraceServerConnectionStatusClient` and would like your input.

File: theia-extensions/viewer-prototype/src/browser/trace-server-connection-status-client-impl.ts
Comment (williamsyang-work) [reply theia-extensions/viewer-prototype/src/browser/trace-server-connection-status-client-impl.ts]: Changed.

**Review by MatthewKhouzam — DISMISSED** [review]
I think you need a second reviewer. Pick @marcdumais-work  or @hriday-panchasara  please.

**Comment by MatthewKhouzam** [comment]
@friends, we need a second review. :)

**Comment by sgraband** [comment]
> @sgraband I'm proposing changes to the TraceServerConnectionStatusClient and would like your input.

No objections from my side. Looks good.

File: theia-extensions/viewer-prototype/src/browser/trace-server-connection-status-client-impl.ts
Comment (williamsyang-work) [inline theia-extensions/viewer-prototype/src/browser/trace-server-connection-status-client-impl.ts:31]: The unused `renderStatus` method has been removed.

**Review by MatthewKhouzam — APPROVED** [review]
Still looks good to me!

**Review by hriday-panchasara — COMMENTED** [review]
LGTM

**Pull request merged**
